Search screen/search screen and pagination

ofekcMoveo opened the pull request:
In this branch I created the search screens - one screen displays the 10 latest searches (saved on the device using userDefualts), and one screen displays the articles we got as a result of the search from the API
 I also implemented pagination in displaying the articles in the tables

ofekcMoveo committed:
add Request struct that represents a generic http request
add structs that contains the JSON objects as recived from the api
change to MVVM-
1. add HomeControllerViewModel,
2. add Repositories folder with ArticleRepository.swift with manages the building and the sending of the request with alamofire,
3. add Networking folder with AlamofireManager.swift which sends a request
4. add methods to HomepageController in order to support the MVVM model and get the data from the modelView

add loader to screen while fetching data from api

ofekcMoveo committed:
add search screen ui add search results iu + VC + viewModels
add functions to ArticleRepository which brings articles by keywords from the api

ofekcMoveo committed:
add pagination
connect actions to back buttons, clear button and remove buttons in the cells

ofekcMoveo committed:
changes in pagination logic

ofekcMoveo committed:
fix readability on the code,
add utils file,
stop animating activityIndicator if there is error

noamkuMoDev commented on Dispatcher/Supporting Files/Utils.swift:
```diff
@@ -0,0 +1,41 @@
+
+struct Utils {
+    
+    static func showErrorAlert(_ msg: String) -> UIAlertController {
```
The function only returns an UIAlertController and doesn't actually display it on the screen so the func name should be accordingly

noamkuMoDev commented on Dispatcher/Repositories/ArticlesRepository.swift:
```diff
@@ -0,0 +1,91 @@
+            method: .get)
+    }
+    
+    func getArticlesByKeywords(_ searchKeywords: String, pageNumber: Int, completionHandler: @escaping ( _ articles: [Article], _ totalPages: Int, _ errorMsg: String?) -> Void) {
```
I think a better name will be something like "getArticlesByUserSearchWords" as its a bit more accurate

noamkuMoDev commented on Dispatcher/Repositories/ArticlesRepository.swift:
```diff
@@ -0,0 +1,91 @@
+    
+private func buildSearchArticlesRequest(_ searchKeywords: String, _ pageNumber: Int) -> Request {
+    return Request(
+        baseUrl: APIConstants.searchArticlesNewscatcherURL,
```
You have 2 methods in this file that build a Request with mostly similar contents (url, headers, method). Better have a dedicated function that does this for you and accepts properties that can be different (e.g. parameters array) only if there are indeed different parameters to pass into it

noamkuMoDev commented on Dispatcher/Constants/APIConstants.swift:
```diff
@@ -0,0 +1,15 @@
+import Foundation
+
+struct APIConstants {
+    static let latestHeadlinesNewscatcherURL = "https://api.newscatcherapi.com/v2/latest_headlines"
```
Basically the baseUrl is the same (https://api.newscatcherapi.com/v2/) and only the endpoint changes. So better have a baseUrl variable and then have different endpoints variables that you stick onto it when needed while you construct the url string when you build the request

noamkuMoDev commented on Dispatcher/Constants/APIConstants.swift:
```diff
@@ -0,0 +1,15 @@
+    static let latestHeadlinesNewscatcherURL = "https://api.newscatcherapi.com/v2/latest_headlines"
+    static let searchArticlesNewscatcherURL = "https://api.newscatcherapi.com/v2/search"
+    static let APIkey = "pxIuyKmcbGplzZfNRXbSQOKltrJ7Zgi4gHsvdWoM43g"
+    static let ARTICLES_TO_FETCH_AMOUNT = 12
```
Stick with the same naming convention - don't have camelCase but then also THIS_WAY and also THISway

noamkuMoDev commented on Dispatcher/Constants/AppConstants.swift:
```diff
@@ -0,0 +1,32 @@
+    
+    
+    //MARK: segues
+    static let fromLatestSearchToResults = "fromLatestSearchToResults"
```
Let's give the segues a dedicated struct of its own, preferably in a separate file as we could potentially have many of them

noamkuMoDev commented on Dispatcher/Constants/AppConstants.swift:
```diff
@@ -0,0 +1,32 @@
+struct AppConstants {
+    //MARK: Global instanses
+    static let articlesRepository = ArticlesRepository()
+    static let defaults = UserDefaults.standard
```
We want to have a dedicated UserDefaultsManager in which we will hold the let defaults = UserDefaults.standard

noamkuMoDev commented on Dispatcher/UI/Search/SearchScreenViewController.swift:
```diff
@@ -0,0 +1,115 @@
+
+    @IBOutlet weak var searchTableView: UITableView!
+    @IBOutlet weak var clearButton: UIButton!
+    @IBOutlet weak var babkButton: UIButton!
```
"backButton" :)

noamkuMoDev commented on Dispatcher/UI/Search/SearchScreenViewController.swift:
```diff
@@ -0,0 +1,115 @@
+        searchTableView.reloadData()
+    }
+    
+    private func customSearchBar() {
```
not an informative function name - we can't know what does the function do unless we read its code

noamkuMoDev commented on Dispatcher/UI/Search/SearchScreenViewController.swift:
```diff
@@ -0,0 +1,115 @@
+    
+    override func viewWillAppear(_ animated: Bool) {
+        searchScreenViewModel.fetchLatestSearchs()
+        searchTableView.reloadData()
```
We want to reload the tableView only after we are sure that the data was fetched as this can fail

noamkuMoDev commented on Dispatcher/Views/SearchTableCells/LatestSearchCell.swift:
```diff
@@ -0,0 +1,40 @@
+        super.awakeFromNib()
+        
+        self.contentView.backgroundColor = UIColor(named: "screenBackgroungColor")
+        removeButton.isHidden = true
```
Why hide the x button and then when creating the cell then you do hide = false?

noamkuMoDev commented on Dispatcher/UI/Homepage/HomepageViewController.swift:
```diff
@@ -6,14 +6,18 @@
-    let articlesToDisplay = Articles().articles
+    let homepageViewModel = HomepageViewModel()
+    let activityIndicator = UIActivityIndicatorView()
+    var currentPage = 1
```
Should be saved inside the viewModel and not the HomepageVC

noamkuMoDev commented on Dispatcher/UI/Homepage/HomepageViewController.swift:
```diff
@@ -23,12 +27,35 @@ class HomeViewController: UIViewController {
+            } else {
+                DispatchQueue.main.async {
+                    self.activityIndicator.stopAnimating()
+                    self.currentPage += 1
```
Should happen inside the viewModel once we know we fetched successfully and added the data to our array

noamkuMoDev commented on Dispatcher/UI/Search/SearchScreenViewController.swift:
```diff
@@ -0,0 +1,115 @@
+        let currentSearchIndex = (searchScreenViewModel.latestSearches.count - indexPath.row) - 1
+        let search = searchScreenViewModel.latestSearches[currentSearchIndex]
+        
+        if let cell = (tableView.dequeueReusableCell(withIdentifier: AppConstants.latestSearchesCellIdentifier, for: indexPath) as? LatestSearchCell) {
```
If we can't register the custom cell that we want we should return a regular UITableViewCell() (like you did in the homepage)

noamkuMoDev commented on Dispatcher/UI/Search/SearchScreenViewController.swift:
```diff
@@ -0,0 +1,115 @@
+extension SearchScreenViewController: UITableViewDelegate, UISearchBarDelegate, LatestSearchCellDelegate {
+    func searchCellSelected(_ search: String) {
+        searchBar.text = search
+        performSegue(withIdentifier: AppConstants.fromLatestSearchToResults, sender: self)
```
We need to display the search results inside the table of the homepage, not navigate to a dedicated screen for showing the results as this creates a confusing UX flow:
user sees general articles -> user searches by words -> user sees wanted results -> user must go back to search screen even though he doesn't want to search anything new -> user sees again general articles

So you need to delete the SearchResultsScreen and display the newly fetched filtered results in the homepage

noamkuMoDev commented on Dispatcher/UI/Search/SearchScreenViewController.swift:
```diff
@@ -0,0 +1,115 @@
+        performSegue(withIdentifier: AppConstants.fromLatestSearchToResults, sender: self)
+    }
+    
+    func removeButtonPressed(_ searchToRemove: String) {
```
We need a better name because when implementing the func here it's not clean enough what it's related to. Something like "removeSearchCellBtnPressed"

noamkuMoDev commented on Dispatcher/UI/Search/SearchScreenViewController.swift:
```diff
@@ -0,0 +1,115 @@
+        self.dismiss(animated: true)
+    }
+    
+    @IBAction func clearButtonPressed(_ sender: UIButton) {
```
in the context of the design of the screen it's not clear whether it's the x button of the searchbar or the square clear button of the saved searches.
So a better name will be "clearAllRecentSearchesBtnPressed" or something like that

noamkuMoDev commented on Dispatcher/UI/Search/SearchViewModel.swift:
```diff
@@ -0,0 +1,66 @@
+
+class SearchViewModel {
+    var articlesToDisplay: [Article] = []
+    var latestSearches: [String] = []
```
Better create a dedicated model for the search object as in the future we might want to add more stuff to it.
So instead of [String] it should be [RecentSearchModel] or so.

noamkuMoDev commented on Dispatcher/UI/Search/SearchViewModel.swift:
```diff
@@ -0,0 +1,66 @@
+    }
+    
+    func fetchLatestSearchs() {
+        latestSearches = (AppConstants.defaults.array(forKey: AppConstants.latestSearchesDefualtsKey) as? [String]) ?? []
```
We want a search repository that will communicate with our UserDefaultsManager to comply with the MVVM+repository architecture that we have in our project

abenartz commented on Dispatcher/UI/Homepage/HomepageViewModel.swift:
```diff
@@ -0,0 +1,44 @@
+    func getTopArticlesFromAPI(pageNumber: Int, completionHandler: @escaping (_ errorMsg: String?) -> Void) {
+        if(isPaginating == false) {
+            isPaginating = true
+            AppConstants.articlesRepository.getLatestArticles(pageNumber: pageNumber) { articles, totalPages,  errorMsg in
```
ArticleRepository (a class) should not be inside AppConstants(file global constant values).
You have two options:
1. Create an instance of the repository inside your ViewModel.
2. Make ArticleRepository singleton.

What is better for our case here?

abenartz commented on Dispatcher/UI/Homepage/HomepageViewModel.swift:
```diff
@@ -0,0 +1,44 @@
+    var isPaginating = false
+     
+    func getTopArticlesFromAPI(pageNumber: Int, completionHandler: @escaping (_ errorMsg: String?) -> Void) {
+        if(isPaginating == false) {
```
Check here also if the paging was exhausted (when you won't get more results)

ofekcMoveo committed:
delete the Utils struct
fix functions names
clean the ArticleResponse object
create const var for tableRowHeight
create "setupTableView" func in homepage controller
fix constraints in ArticleCell
organize files in groups

ofekcMoveo committed:
merge branch Alamofire to branch Search_Screen_and_Pagination

ofekcMoveo committed:
solve conflicts in project file

ofekcMoveo committed:
add colorPalleteNames struct
add reading api key from json file in the AppDelegate file
add .gitignore file
little changes in the controllers and view models

ofekcMoveo committed:
make homepage view model a singleton

ofekcMoveo committed:
add extension to String in order to send string as error description
change name of buttons in searchScreen and latestSearchCell
make ArticlesRepository a singleton

ofekcMoveo committed:
add images to assets
add model RecentSearches

ofekcMoveo committed:
add a search repository

ofekcMoveo committed:
delete an unnecessary decleration

noamkuMoDev commented on Dispatcher/Supporting Files/Utils.swift:
```diff
@@ -0,0 +1,61 @@
+}
+
+
+func createErrorAlert(_ msg: String) -> UIAlertController {
```
* The naming convention would be to use "getErrorAlert" rather than "createErrorAlert"
* It would be nicer to read when calling this function if the signature was: getErrorAlert(withMessage msg: String)

noamkuMoDev commented on Dispatcher/Supporting Files/AppDelegate.swift:
```diff
@@ -14,6 +14,13 @@ class AppDelegate: UIResponder, UIApplicationDelegate {
 
 
     func application(_ application: UIApplication, didFinishLaunchingWithOptions launchOptions: [UIApplication.LaunchOptionsKey: Any]?) -> Bool {
+        let response = readApiKeyFromConfigFile()
```
Instead of "response" a better naming would be what it actually is - apiKeyValue

noamkuMoDev commented on Dispatcher/Supporting Files/Config.json:
```diff
@@ -0,0 +1,3 @@
+{
+    "api_key": "oE0GNhaEFfOy9eAygqEaq4P9ir8hqxrIui7u20H5_e4"
```
Need to add the Config.json to the .gitignore file so that it wouldn't be uploaded to github

noamkuMoDev commented on Dispatcher/Repositories/ArticlesRepository.swift:
```diff
@@ -0,0 +1,89 @@
+    }
+    
+    private func buildRequest(_ searchKeywords: String? , _ pageNumber: Int) -> Request {
+        if(searchKeywords != nil) {
```
When you need to use the variable that you are checking inside the { } it's better to use "if let" instead of "!= nil" because then you don't need to force-unwrap the variable using "!".
Even though it won't fail because you indeed checked it's not null, we still don't like to use the "!"

noamkuMoDev commented on Dispatcher/Constants/AppConstants.swift:
```diff
@@ -0,0 +1,40 @@
+    
+    //MARK: Homepage articles table
+    static let tableRowHight = 40
+    static let topHeadlinesHeaderText = "Top Headlines in Israel"
```
Better have a separate file for text constants

noamkuMoDev commented on Dispatcher/Constants/AppConstants.swift:
```diff
@@ -0,0 +1,40 @@
+    
+    //MARK: Search
+    static let latestSearchesAmount = 10
+    static let latestSearchesDefualtsKey = "LatestSearches"
```
Better have a separate struct for UserDefaultsKeys

noamkuMoDev commented on Dispatcher/Networking/AlamofireManager.swift:
```diff
@@ -0,0 +1,29 @@
+class AlamofireManager {
+    
+    func sendRequest<T:Decodable>(_ request: Request, completionHandler: @escaping (Result<T, Error>) -> Void) {
+        if let encoded = request.urlWithParams.addingPercentEncoding(withAllowedCharacters: .urlFragmentAllowed),let url = URL(string: encoded) {
```
What happens in case the "if let" fails?

noamkuMoDev commented on Dispatcher/Supporting Files/StringErrorExtension.swift:
```diff
@@ -0,0 +1,12 @@
+
+import Foundation
+
+extension String: LocalizedError {
```
Why did you need to add this?

noamkuMoDev commented on Dispatcher/UI/Homepage/HomepageViewModel.swift:
```diff
@@ -0,0 +1,50 @@
+    private init () {}
+     
+    func getArticlesToDisplay(completionHandler: @escaping (_ errorMsg: String?) -> Void) {
+        if(isPaginating == false && currentPage <= totalResultsPages) {
```
Better convention is to check for false by: !isPagination

noamkuMoDev commented on Dispatcher/UI/Homepage/HomepageViewModel.swift:
```diff
@@ -0,0 +1,50 @@
+            isPaginating = true
+            ArticlesRepository.shared.getArticlesFromApi(pageNumber: currentPage) { articles, totalPages,  errorMsg in
+                if (errorMsg != nil) {
+                    completionHandler(errorMsg!)
```
If we failed then we never reset the isPaginating back to false?

noamkuMoDev commented on Dispatcher/Repositories/ArticlesRepository.swift:
```diff
@@ -0,0 +1,89 @@
+    private init() {}
+
+    func getArticlesFromApi(pageNumber: Int, completionHandler: @escaping (_ articles: [Article], _ totalPages: Int, _ errorMsg: String?) -> Void) {
+        let request = buildRequest(nil, pageNumber)
```
When you don't provide names for the parameters that you send in it's much less readable and harder to understand. You actually have to go into the buildRequest function itself to understand that in this particular case you don't have search words and that's why you send nil into it.

noamkuMoDev commented on Dispatcher/Repositories/SearchRepository.swift:
```diff
@@ -0,0 +1,34 @@
+
+class SearchRepository {
+    
+    static let shared = SearchRepository()
```
Why did you create it as a singleton?

noamkuMoDev commented on Dispatcher/Persistence/UserDefaultsManager.swift:
```diff
@@ -0,0 +1,24 @@
+//
+//  UserDefaultsManager.swift
```
Since this is a Manager file, we would want the functions inside it to be as general as possible (like in the AlamofireManager) so we can use it wherever we need throughout the app.
We don't want to have to create a dedicated function for every data that we will need to save to the user defaults (e.g. if the user saw onboarding or not - will we create a "saveSeenOnboarding" function as well?)

noamkuMoDev commented on Dispatcher/UI/Homepage/HomepageViewController.swift:
```diff
@@ -0,0 +1,145 @@
+        let currentArticle = homepageViewModel.getArticleByIndex(index: indexPath.row)
+        
+        if let cell = (tableView.dequeueReusableCell(withIdentifier: TableCellsIdentifiers.articleCellIdentifier, for: indexPath) as? ArticleCell) {
+            DispatchQueue.main.async {
```
The cellForRowAt is a synchronic function that is always executed on the main thread automatically

noamkuMoDev commented on Dispatcher/UI/Homepage/HomepageViewController.swift:
```diff
@@ -0,0 +1,145 @@
+    func tableView(_ tableView: UITableView, willDisplay cell: UITableViewCell, forRowAt indexPath: IndexPath) {
+        if indexPath.row == (homepageViewModel.articlesToDisplay.count - 2) {
+            if (homepageViewModel.currentPage < homepageViewModel.totalResultsPages) {
+                getArticles()
```
Eventually we reload the entire tableView just because we added a few more articles (possibly even just 1) to our existing data array (could already have even 100 items).
This is wasteful in resources.
Instead need to use insertRows to just add the new rows to the tableview and leave the existing rows untouched

noamkuMoDev commented on Dispatcher/Repositories/ArticlesRepository.swift:
```diff
@@ -0,0 +1,89 @@
+        }
+    }
+    
+    func getArticlesByUserSearchWords(_ searchKeywords: String, pageNumber: Int, completionHandler: @escaping ( _ articles: [Article], _ totalPages: Int, _ errorMsg: String?) -> Void) {
```
getArticlesByUserSearchWords & getArticlesFromApi are pretty much exactly the same aside from the searchKeywords. So no need for 2 functions

noamkuMoDev commented on Dispatcher/Constants/APIConstants.swift:
```diff
@@ -0,0 +1,15 @@
+
+struct APIConstants {
+    static var apiKey: String = ""
+    static let latestHeadlinesNewscatcherURL = "https://api.newscatcherapi.com/v2"
```
This is the same url for both variables.
Call it "baseUrl" instead.
And save the different endpoints

noamkuMoDev commented on Dispatcher/Repositories/SearchRepository.swift:
```diff
@@ -0,0 +1,34 @@
+
+class SearchRepository {
+    
+    static let shared = SearchRepository()
```
Why is it a singleton? Only the SeachViewModel uses it

ofekcMoveo replied:
I did it a singleton because the ArticlesRepository is a singleton and I thought that it should be the same in order to be consistent

noamkuMoDev replied:
הסינגלטון הוא פיתרון שימושי למצבים שבהם אנחנו חייבים שהקלאס יהיה זמין ונגיש לנו ברחבי האפליקציה והדאטה שהוא מחזיק יהיה זהה מכל מקום שניגש אליו - זה עונה על קונספט של single source of truth
אבל חיסרון מרכזי של הסינגלטון הוא שברגע שהוא נוצר לראשונה - הוא יחייה לנצח כל עוד האפליקציה בחיים, גם אם כבר אין לנו בו צורך או שימוש, והוא ממשיך לגזול משאבים של זיכרון שעולה לנו בפרפורמנס של האפליקציה.
ולכן אנחנו לא ניצור סתם סינגלטונים, אלא רק אם יש באמת צורך בכך. במקרה פה למשל, רק ה-SearchViewModel צריך להשתמש ב-SearchRepository ולכן אין לנו צורך ליצור ממנו סינגלטון

ofekcMoveo committed:
Add .gitignore file

ofekcMoveo committed:
change UserDefualts manager to be more general
change reloading to the UItableview to insert rows
add TextConstants file
add Errors enum

ofekcMoveo committed:
make searchRepository not a singleton
